### ISSUE TITLE
Invalid memory access when sending response to closed stream

### ISSUE THREAD

**Issue by cellscape** [tatsuhiro-t/nghttp2#312]
When sending response to already closed stream, nghttp2 tries to access already freed `stream->item` and then tries to free it second time.

I've triggered the issue by incorrectly sending response headers to just reset server stream. I understand it's an error from my side but I believe nghttp2 still shouldn't corrupt memory in such case, only report error as it does already.

My understanding is that item attached to stream is not properly detached but only freed when failed to send frame to already closed stream. And then it is double freed in `nghttp2_session_del`.

Below is the output from nghttp2 debug and gcc address sanitizer. I had to remove frames from my application but there still should be enough information to understand what's going on. Please let me know if you need more information, I can easily reproduce the issue if needed.

```
recv: connection recv_window_size=0, local_window=65535
recv: [IB_READ_HEAD]
recv: payloadlen=4, type=3, flags=0x00, stream_id=13
recv: RST_STREAM
recv: [IB_READ_NBYTE]
recv: readlen=4, payloadleft=0, left=0
stream: stream=13 attach item=0x60c000016d80
stream: stream=13 enqueue
stream: keep closed stream(0x610000006440)=13, state=1
stream: adjusting kept closed streams num_closed_streams=1, num_incoming_streams=0, max_concurrent_streams=100
send: frame preparation failed with Stream was already closed or invalid
send: reset nghttp2_active_outbound_item
send: aob->item = (nil)
recv: connection recv_window_size=0, local_window=65535
recv: [IB_READ_HEAD]
recv: payloadlen=8, type=7, flags=0x00, stream_id=0
recv: GOAWAY
recv: [IB_READ_NBYTE]
recv: readlen=8, payloadleft=0, left=0
recv: [IB_READ_GOAWAY_DEBUG]
recv: readlen=0, payloadleft=0

==25039==ERROR: AddressSanitizer: heap-use-after-free on address 0x60c000016df0 at pc 0x5b3313 bp 0x7fff94d73530 sp 0x7fff94d73520
READ of size 1 at 0x60c000016df0 thread T0 (main)
    #0 0x5b3312 in free_streams (main+0x5b3312)
    #1 0x5e3420 in nghttp2_map_each_free (main+0x5e3420)
    #2 0x5b3546 in nghttp2_session_del (main+0x5b3546)
    #3 0x4f23a4 in ...

0x60c000016df0 is located 112 bytes inside of 120-byte region [0x60c000016d80,0x60c000016df8)
freed by thread T0 (main) here:
    #0 0x7f555d8b948f in __interceptor_free (/usr/lib/gcc/x86_64-pc-linux-gnu/4.9.3/libasan.so.1+0x5748f)
    #1 0x5df711 in default_free (main+0x5df711)
    #2 0x5df854 in nghttp2_mem_free (main+0x5df854)
    #3 0x5bb641 in nghttp2_session_mem_send_internal (main+0x5bb641)
    #4 0x5bc3a5 in nghttp2_session_send (main+0x5bc3a5)
    #5 0x4f213a in ...

previously allocated by thread T0 (main) here:
    #0 0x7f555d8b9707 in malloc (/usr/lib/gcc/x86_64-pc-linux-gnu/4.9.3/libasan.so.1+0x57707)
    #1 0x5df6f3 in default_malloc (main+0x5df6f3)
    #2 0x5df7e1 in nghttp2_mem_malloc (main+0x5df7e1)
    #3 0x5ccf07 in submit_headers_shared (main+0x5ccf07)
    #4 0x5cd490 in submit_headers_shared_nva (main+0x5cd490)
    #5 0x5ce21a in nghttp2_submit_response (main+0x5ce21a)
    ...
    #16 0x5b561d in nghttp2_session_close_stream (main+0x5b561d)
    #17 0x5bf638 in nghttp2_session_on_rst_stream_received (main+0x5bf638)
    #18 0x5bf714 in session_process_rst_stream_frame (main+0x5bf714)
    #19 0x5c6f46 in nghttp2_session_mem_recv (main+0x5c6f46)
    #20 0x5c9fc7 in nghttp2_session_recv (main+0x5c9fc7)
```

**Comment by tatsuhiro-t** [comment]
Thank you for reporting this.  We handled detachment of item when stream was closed, but the reported case is really edge case.  I fixed this bug in e66bd49.

**Comment by cellscape** [comment]
Thanks, fix works fine for us.
